internal/app/starter/master_linux.go

```diff
@@ -11,129 +11,131 @@ import (
-		} else {
-			rpcConn.Close()
-		}
+func startContainer(masterSocket int, containerPid int, e *engine.Engine, fatalChan chan error) {
+	comm := os.NewFile(uintptr(masterSocket), "master-socket")
+	if comm == nil {
+		fatalChan <- fmt.Errorf("bad master socket file descriptor")
+		return
+	}
+	conn, err := net.FileConn(comm)
+	comm.Close()
+	if err != nil {
+		fatalChan <- fmt.Errorf("failed to create master connection: %s", err)
+		return
+	}
+	defer conn.Close()
 
-		runtime.Goexit()
-	}()
+	data := make([]byte, 1)
 
-	go func() {
-		data := make([]byte, 1)
-		comm := os.NewFile(uintptr(masterSocket), "master-socket")
-		conn, err := net.FileConn(comm)
-		comm.Close()
+	// special path for engines which needs to stop before executing
+	// container process
+	if obj, ok := e.Operations.(interface {
+		PreStartProcess(int, net.Conn, chan error) error
+	}); ok {
+		_, err := conn.Read(data)
 		if err != nil {
-			fatalChan <- fmt.Errorf("failed to create master connection: %s", err)
-		}
-		defer conn.Close()
-
-		// special path for engines which needs to stop before executing
-		// container process
-		if obj, ok := e.Operations.(interface {
-			PreStartProcess(int, net.Conn, chan error) error
-		}); ok {
-			_, err := conn.Read(data)
-			if err != nil {
-				if err != io.EOF {
-					fatalChan <- fmt.Errorf("error while reading master socket data: %s", err)
-				}
-				// EOF means something goes wrong in stage 2, don't send error via
-				// fatalChan, error will be reported by stage 2 and the process
-				// status will be set accordingly via MonitorContainer method below
-				sylog.Debugf("stage 2 process was interrupted, waiting status")
-				return
-			} else if data[0] == 'f' {
-				// StartProcess reported an error in stage 2, don't send error via
-				// fatalChan, error will be reported by stage 2 and the process
-				// status will be set accordingly via MonitorContainer method below
-				sylog.Debugf("stage 2 process reported an error, waiting status")
-				return
-			}
-			if err := obj.PreStartProcess(containerPid, conn, fatalChan); err != nil {
-				fatalChan <- fmt.Errorf("pre start process failed: %s", err)
+
+			if err != io.EOF {
+				fatalChan <- fmt.Errorf("error while reading master socket data: %s", err)
 				return
 			}
-		}
-		// wait container process execution, EOF means container process
-		// was executed and master socket was closed by stage 2. If data
-		// byte sent is equal to 'f', it means an error occurred in
-		// StartProcess, just return by waiting error and process status
-		_, err = conn.Read(data)
-		if (err != nil && err != io.EOF) || data[0] == 'f' {
+			// EOF means something goes wrong in stage 2, don't send error via
+			// fatalChan, error will be reported by stage 2 and the process
+			// status will be set accordingly via MonitorContainer method below
+			sylog.Debugf("stage 2 process was interrupted, waiting status")
+			return
+		} else if data[0] == 'f' {
```

Comment:
Can be simplified to `if data[0] == 'f'` since you have a return statement in previous if clause

Reply:
The linter did not complain about this, I believe this is totally minor based on personal preferences.